File: python_code/common_codes/a5_Fit_lm_model.py
import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import KFold
from sklearn.metrics import r2_score
import statsmodels.api as sm


def a5_Fit_lm_model(Pred, Resp):
    Resp = Resp.ravel()
    columnNames = ['Model', 'R2', 'adj-R2', 'AIC', 'R2CV']
    performance = pd.DataFrame(columns=columnNames)

    X = sm.add_constant(Pred)
    lm_sm = sm.OLS(Resp, X).fit()
    R2 = lm_sm.rsquared
    adj_R2 = lm_sm.rsquared_adj
    AIC = lm_sm.aic

    kf = KFold(n_splits=5, shuffle=True, random_state=1)
    yPredCV = np.zeros(len(Resp))
    for train_idx, test_idx in kf.split(Pred):
        X_train, X_test = Pred[train_idx], Pred[test_idx]
        y_train = Resp[train_idx]
        mdl_cv = LinearRegression().fit(X_train, y_train)
        yPredCV[test_idx] = mdl_cv.predict(X_test).ravel()

    SSE = np.sum((yPredCV - Resp)**2)
    SST = np.sum((np.mean(Resp) - Resp)**2)
    R2CV = 1 - (SSE / SST)

    performance.loc[0] = ['linear', R2, adj_R2, AIC, R2CV]

    Modells_flm = LinearRegression()
    Modells_flm.fit(Pred, Resp)

    return Modells_flm, performance

File: common_codes/a5_Fit_lm_model.m
function a5_Fit_lm_model(Pred, Resp)
%Pred=normalize(Pred,'center','mean');
%tb=coding(Pred);

type={'linear','interactions'};
columnNames = {'Model', 'DF', 'SumSqr','F-value','p-value','R2','adj-R2','AIC','AICc','R2CV'};
performance = table('Size', [0, numel(columnNames)], 'VariableNames', columnNames,...
    'VariableTypes',['string',repmat({'double'},1,9)]);
for i=1:1
    Modells_flm=fitlm(Pred,Resp,type{i});
    R2CV=crosst(Pred,Resp,type{i});
    tab=anova(Modells_flm,'summary');
    res={type{i},tab.DF(2),tab.SumSq(2),tab.F(2),tab.pValue(2),Modells_flm.Rsquared.Ordinary,...
        Modells_flm.Rsquared.Adjusted,Modells_flm.ModelCriterion.AIC,Modells_flm.ModelCriterion.AICc,R2CV};
    performance(i,:)=res;
end

save Fitlm_model.mat Modells_flm performance
disp('');



function R2CV=crosst(Pred,Yi,mtype)
cv = cvpartition(height(Pred),'KFold',5);
yPredCV = zeros(height(Pred),1);
for k = 1:cv.NumTestSets
    trainIdx = training(cv,k);
    testIdx = test(cv,k);
    mdl_cv = fitlm(Pred(trainIdx,:), Yi(trainIdx,:), mtype);
    yPredCV(testIdx) = predict(mdl_cv, Pred(testIdx, :));
end
SSE=sum((yPredCV-Yi).^2);
SST=sum((mean(Yi)-Yi).^2);
R2CV=1-(SSE/SST);

function coded=coding(tab)
format short
data=tab(:,:);
mini=min(data,[],1);
maxi=max(data,[],1);
data=(2*((data-mini)./(maxi-mini))-1)*1.633;
coded=round(data,3);

